spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1349,11 +1363,11 @@ private void commitIfNecessary() {
 			return commits;
 		}
 
-		private Collection<ConsumerRecord<K, V>> getHighestOffsetRecords(List<ConsumerRecord<K, V>> records) {
+		private Collection<ConsumerRecord<K, V>> getHighestOffsetRecords(ConsumerRecords<K, V> records) {
 			final Map<TopicPartition, ConsumerRecord<K, V>> highestOffsetMap = new HashMap<>();
-			records.forEach(r -> {
-				highestOffsetMap.compute(new TopicPartition(r.topic(), r.partition()),
-						(k, v) -> v == null ? r : r.offset() > v.offset() ? r : v);
+			records.partitions().forEach(tp -> {
```

Comment:
`records.partitions().stream()...collect(Collectors.toMap())` instead?

Reply:
I don't think that will be efficient because we'd have to iterate over the lists to get the highest rather than indexing directly.

Reply:
Correction - I figured it out.